File: examples/docs-code-snippets/src/main/kotlin/examples/kotlin/interactions/CheckStockImpl.kt
package examples.kotlin.interactions

object CheckStockImpl : CheckStock {
    override fun apply(orderId: String, productIds: List<String>): CheckStock.Outcome {
        println("Checking stock for order: $orderId and products: $productIds")

        return if ((1..1000).random() < 500) {
            CheckStock.SufficientStock
        } else {
            CheckStock.OrderHasUnavailableItems(productIds)
        }
    }
}

File: examples/docs-code-snippets/src/main/scala/examples/scala/interactions/CheckStockImpl.scala
package examples.scala.interactions

import scala.concurrent.Future
import scala.util.Random

trait CheckStockTrait {
  def apply(orderId: String, productIds: List[String]): Future[CheckStock.Outcome]
}

class CheckStockImpl extends CheckStockTrait {
  override def apply(orderId: String, productIds: List[String]): Future[CheckStock.Outcome] = {
    println(s"Checking stock for order: $orderId and products: $productIds")

    val randomNumber = new Random().nextInt(1000) + 1
    if (randomNumber < 500) {
      Future.successful(CheckStock.SufficientStock())
    } else {
      Future.successful(CheckStock.OrderHasUnavailableItems(productIds))
    }
  }
}
